src/adaptors/adaptor_tls.c

```diff
@@ -0,0 +1,630 @@
+{
+    pn_raw_buffer_t take_incoming_buf;
+    size_t          take_input_count;
+    while ((take_input_count = pn_tls_take_decrypt_input_buffers(tls->tls_session, &take_incoming_buf, 1))) {
+        assert(take_input_count == 1);
+        (void) take_input_count;  // prevent unused variable warning
+        qd_adaptor_buffer_t *adaptor_buff = (qd_adaptor_buffer_t *) take_incoming_buf.context;
+        qd_adaptor_buffer_free(adaptor_buff);
+    }
+}
+
+size_t get_pn_raw_buffer_fetch_size(qd_tls_t *tls)
+{
+    size_t tls_input_buff_capacity = pn_tls_get_decrypt_input_buffer_capacity(tls->tls_session);
+    size_t fetch_size              = MIN(tls_input_buff_capacity, TLS_MAX_INPUT_CAPACITY);
+    return fetch_size;
+}
+
+int qd_tls_decrypt(qd_tls_t *tls, pn_raw_connection_t *pn_raw_conn, qd_adaptor_buffer_list_t *decrypted_buffs)
+{
+    if (tls->tls_error)
+        return QD_TLS_ERROR;
+
+    int    encrypted_bytes_in = 0;
+    size_t fetch_size         = get_pn_raw_buffer_fetch_size(tls);
+
+    // Do not proceed if TLS has no decrypt input capacity.
+    if (fetch_size == 0) {
+        qd_log(tls->log_source, QD_LOG_TRACE, "[C%" PRIu64 "] qd_tls_decrypt fetch_size=0, returning", tls->conn_id);
+        return 0;
+    }
+
+    pn_raw_buffer_t raw_buffers[TLS_MAX_INPUT_CAPACITY];
+    size_t          n;
+    while ((n = pn_raw_connection_take_read_buffers(
+                pn_raw_conn,
+                raw_buffers,
+                fetch_size))) {  // Repeatedly get as much buffers (from raw connection) as possible
+        for (size_t i = 0; i < n; ++i) {
+            // Give each decrypt input buffer to proton tls
+            qd_adaptor_buffer_t *adaptor_buff = (qd_adaptor_buffer_t *) raw_buffers[i].context;
+            if (raw_buffers[i].size > 0) {
+                encrypted_bytes_in += raw_buffers[i].size;
+                size_t consumed = pn_tls_give_decrypt_input_buffers(tls->tls_session, &raw_buffers[i], 1);
+                qd_log(tls->log_source,
+                       QD_LOG_TRACE,
+                       "[C%" PRIu64 "] qd_tls_decrypt gave raw buffer to pn_tls_give_decrypt_input_buffers, size=%zu ",
+                       tls->conn_id,
+                       raw_buffers[i].size);
+                (void) consumed;  // prevent unused variable warning
+                assert(consumed == 1);
+            } else {
+                //
+                // This buffer size is zero, we will free this buffer immediately.
+                //
+                qd_log(tls->log_source,
+                       QD_LOG_TRACE,
+                       "[C%" PRIu64 "] qd_tls_decrypt raw buffer size=0, did not sent this buffer to TLS, freeing it",
+                       tls->conn_id);
+                qd_adaptor_buffer_free(adaptor_buff);
+            }
+        }
+
+        //
+        // Process TLS.
+        //
+        if (!process_tls(tls)) {
+            take_back_input_decrypt_buffs(tls);
```

Comment:
Shouldn't take_back_input_decrypt_buffs() be called at the start of this function in order to make room for adding new input decrypt buffers?

Reply:
At the end of each call to qd_tls_decrypt() we always take back all or most of the input buffers. If proton does not return an input buffer it means that it does not have all the data that it needs to do a successful decryption. When more data comes in the next time, that originally held buffer is returned when there is enough data to decrypt. So, this should be ok.